reader.go

```diff
@@ -30,33 +30,133 @@ func (o *ReaderOptions) norm(name string) {
 
 // Reader reads data from a remote feed.
 type Reader struct {
-	remote *bfs.Object
-	opt    ReaderOptions
-	ctx    context.Context
-	num    int
+	ctx context.Context
+	opt *ReaderOptions
 
-	br io.ReadCloser // bfs reader
-	cr io.ReadCloser // compression reader
-	fd FormatDecoder
+	remotes  []*bfs.Object
+	cur      *streamReader
+	pos, num int
 }
 
 // NewReader inits a new reader.
 func NewReader(ctx context.Context, remote *bfs.Object, opt *ReaderOptions) (*Reader, error) {
-	var o ReaderOptions
-	if opt != nil {
-		o = *opt
-	}
-	o.norm(remote.Name())
+	return &Reader{
+		remotes: []*bfs.Object{remote},
+		opt:     opt,
+		ctx:     ctx,
+	}, nil
+}
 
+// MultiReader inits a new reader for multiple remotes.  Remotes are read sequentially as if concatenated.
+// Once all remotes are fully read, Read will return EOF.
+func MultiReader(ctx context.Context, remotes []*bfs.Object, opt *ReaderOptions) (*Reader, error) {
 	return &Reader{
-		remote: remote,
-		opt:    o,
-		ctx:    ctx,
+		remotes: remotes,
+		opt:     opt,
+		ctx:     ctx,
 	}, nil
 }
 
 // Read reads raw bytes from the feed.
 func (r *Reader) Read(p []byte) (int, error) {
+	if len(r.remotes) == 0 || r.pos >= len(r.remotes) {
```

Comment:
you are doing this several times, don't you want to do something like `if !r.ensureCurrent() ...`